File: backend/worker/dependency_resolver.py
"""
Dependency Resolution Engine for Python
Handles building dependency graphs, detecting parallel opportunities, and resolving execution groups
"""

import logging
from typing import Dict, List, Set, Tuple, Optional, Any
from enum import Enum
from utils.step_utils import coerce_dependency_value, normalize_dependency_list, normalize_step_order

logger = logging.getLogger(__name__)


class StepStatus(Enum):
    COMPLETED = "completed"
    RUNNING = "running"
    WAITING = "waiting"
    READY = "ready"


# Helper functions

def _build_order_to_index(steps: List[Dict]) -> Dict[int, int]:
    """
    Build a mapping from step_order to array index for normalization.
    
    Args:
        steps: List of step dictionaries
        
    Returns:
        Dictionary mapping step_order to array index
    """
    order_to_index: Dict[int, int] = {}
    for index, step in enumerate(steps):
        step_order = normalize_step_order(step)
        order_to_index[step_order] = index
    return order_to_index


def _normalize_dependency_index(
    dep_value: Any,
    order_to_index: Dict[int, int],
    steps_length: int,
    current_index: int
) -> Optional[int]:
    """
    Normalize a dependency value (step_order or array index) to an array index.
    
    Args:
        dep_value: Dependency value (could be step_order or array index)
        order_to_index: Mapping from step_order to array index
        steps_length: Total number of steps
        current_index: Index of the current step (to prevent self-dependencies)
        
    Returns:
        Normalized array index, or None if invalid
    """
    dep_index_value, _ = coerce_dependency_value(dep_value)
    if dep_index_value is None:
        return None
    
    # Try to normalize: if dep_value matches a step_order, convert to array index
    dep_index = dep_index_value
    if dep_index_value in order_to_index:
        dep_index = order_to_index[dep_index_value]
    elif dep_index_value >= 0 and dep_index_value < steps_length:
        # Already an array index, use as-is
        dep_index = dep_index_value
    else:
        # Invalid - doesn't match any step_order or array index
        return None
    
    # Validate normalized index
    if dep_index < 0 or dep_index >= steps_length or dep_index == current_index:
        return None
    
    return dep_index


def _get_step_dependencies(step: Dict, step_index: int, steps: List[Dict]) -> List[int]:
    """
    Extract dependencies for a step, handling both explicit and auto-detected dependencies.
    
    Args:
        step: Step dictionary
        step_index: Index of the step in the steps list
        steps: List of all step dictionaries
        
    Returns:
        List of dependency indices
    """
    deps: List[int] = []
    
    if step.get('depends_on') and isinstance(step.get('depends_on'), list):
        # Explicit dependencies provided
        deps = normalize_dependency_list(step['depends_on'])
    else:
        # Auto-detect from step_order
        step_order = normalize_step_order(step)
        for i, s in enumerate(steps):
            other_order = normalize_step_order(s)
            if other_order < step_order:
                deps.append(i)
    
    return deps


def build_dependency_graph(steps: List[Dict]) -> Dict[int, List[int]]:
    """
    Build dependency graph from workflow steps.
    
    Args:
        steps: List of step dictionaries
        
    Returns:
        Dictionary mapping step index to list of dependency indices
    """
    dependencies: Dict[int, List[int]] = {}
    order_to_index = _build_order_to_index(steps)
    
    for index, step in enumerate(steps):
        deps: List[int] = []
        
        if step.get('depends_on') and isinstance(step.get('depends_on'), list):
            # Explicit dependencies provided - normalize step_order values to array indices
            for dep_value in normalize_dependency_list(step['depends_on']):
                dep_index = _normalize_dependency_index(
                    dep_value, order_to_index, len(steps), index
                )
                if dep_index is not None:
                    deps.append(dep_index)
        else:
            # Auto-detect from step_order
            step_order = normalize_step_order(step)
            for i, s in enumerate(steps):
                other_order = normalize_step_order(s)
                if other_order < step_order:
                    deps.append(i)
        
        dependencies[index] = deps
    
    return dependencies


def detect_parallel_opportunities(steps: List[Dict]) -> Dict[int, List[int]]:
    """
    Detect parallel opportunities from step_order.
    Steps with the same step_order can run in parallel.
    
    Args:
        steps: List of step dictionaries
        
    Returns:
        Dictionary mapping step_order to list of step indices
    """
    order_groups: Dict[int, List[int]] = {}
    
    for index, step in enumerate(steps):
        step_order = step.get('step_order', index)
        
        if step_order not in order_groups:
            order_groups[step_order] = []
        order_groups[step_order].append(index)
    
    return order_groups


def resolve_execution_groups(steps: List[Dict]) -> Dict:
    """
    Resolve execution groups - group steps into batches that can run in parallel.
    
    Args:
        steps: List of step dictionaries
        
    Returns:
        Dictionary with executionGroups and totalSteps
    """
    if not steps:
        return {
            'executionGroups': [],
            'totalSteps': 0,
        }
    
    dependencies = build_dependency_graph(steps)
    execution_groups: List[Dict] = []
    completed: Set[int] = set()
    group_index = 0
    
    while len(completed) < len(steps):
        # Find all steps that are ready to execute (all dependencies completed)
        ready_steps: List[int] = []
        
        for i in range(len(steps)):
            if i in completed:
                continue  # Already completed
            
            deps = dependencies.get(i, [])
            all_deps_completed = len(deps) == 0 or all(dep_index in completed for dep_index in deps)
            
            if all_deps_completed:
                ready_steps.append(i)
        
        if len(ready_steps) == 0:
            # This shouldn't happen if dependencies are valid, but handle gracefully
            logger.warning(
                f"No ready steps found, but {len(completed)}/{len(steps)} steps completed. "
                "Possible circular dependency."
            )
            break
        
        # Check if steps in this group can run in parallel
        # Steps can run in parallel if they don't depend on each other
        can_run_in_parallel = len(ready_steps) > 1 and not _has_internal_dependencies(ready_steps, dependencies)
        
        execution_groups.append({
            'groupIndex': group_index,
            'stepIndices': ready_steps,
            'canRunInParallel': can_run_in_parallel or len(ready_steps) == 1,
        })
        
        # Mark these steps as completed for next iteration
        completed.update(ready_steps)
        group_index += 1
    
    return {
        'executionGroups': execution_groups,
        'totalSteps': len(steps),
    }


def _has_internal_dependencies(step_indices: List[int], dependencies: Dict[int, List[int]]) -> bool:
    """Check if steps in a group have internal dependencies (can't run in parallel)."""
    for step_index in step_indices:
        deps = dependencies.get(step_index, [])
        # If any dependency is also in this group, they can't run in parallel
        if any(dep_index in step_indices for dep_index in deps):
            return True
    return False


def validate_dependencies(steps: List[Dict]) -> Tuple[bool, List[str]]:
    """
    Validate dependencies - check for circular dependencies and invalid references.
    
    Args:
        steps: List of step dictionaries
        
    Returns:
        Tuple of (is_valid, list_of_errors)
    """
    errors: List[str] = []
    
    if not steps:
        return True, []
    
    order_to_index = _build_order_to_index(steps)
    
    # Check for invalid dependency indices
    coerced_values: List[Dict[str, Any]] = []
    for index, step in enumerate(steps):
        if step.get('depends_on') and isinstance(step.get('depends_on'), list):
            step_order = normalize_step_order(step)
            step_name = step.get('step_name', 'Unknown')
            
            for dep_value in step['depends_on']:
                dep_index_value, was_coerced = coerce_dependency_value(dep_value)
                if dep_index_value is None:
                    errors.append(
                        f"Step {step_order} ({step_name}): "
                        f"depends_on contains invalid value {dep_value} "
                        f"(type={type(dep_value).__name__})"
                    )
                    continue
                if was_coerced:
                    coerced_values.append({
                        "step_order": step_order,
                        "step_name": step_name,
                        "raw_value": dep_value,
                        "normalized": dep_index_value,
                    })
                
                dep_index = _normalize_dependency_index(
                    dep_index_value, order_to_index, len(steps), index
                )
                
                if dep_index is None:
                    # Provide more helpful error message
                    # Check if dep_value matches a step_order
                    matching_order = None
                    for order, idx in order_to_index.items():
                        if order == dep_index_value:
                            matching_order = order
                            break
                    
                    if matching_order is not None:
                        # It matched a step_order but normalization failed (self-dependency or invalid)
                        errors.append(
                            f"Step {step_order} ({step_name}): "
                            f"depends_on step_order {dep_index_value} is invalid "
                            f"(could be self-dependency or step doesn't exist)"
                        )
                    elif dep_index_value >= 0 and dep_index_value < len(steps):
                        # It's a valid array index but normalization failed
                        errors.append(
                            f"Step {step_order} ({step_name}): "
                            f"depends_on array index {dep_index_value} is invalid "
                            f"(could be self-dependency)"
                        )
                    else:
                        # Completely out of range
                        errors.append(
                            f"Step {step_order} ({step_name}): "
                            f"depends_on value {dep_index_value} is out of range "
                            f"(valid step_orders: {sorted(order_to_index.keys())}, "
                            f"valid array indices: 0-{len(steps)-1})"
                        )
                    continue
                
                # Additional validation for error messages
                if dep_index == index:
                    errors.append(
                        f"Step {step_order} ({step_name}): "
                        "cannot depend on itself"
                    )
    
    # Check for circular dependencies using DFS
    visited: Set[int] = set()
    rec_stack: Set[int] = set()
    order_to_index_for_cycle = _build_order_to_index(steps)
    
    def has_cycle(node_index: int) -> bool:
        if node_index < 0 or node_index >= len(steps):
            return False  # Invalid index, skip
        if node_index in rec_stack:
            return True  # Found a cycle
        if node_index in visited:
            return False  # Already processed
        
        visited.add(node_index)
        rec_stack.add(node_index)
        
        step = steps[node_index]
        dep_values = _get_step_dependencies(step, node_index, steps)
        
        # Normalize dependency values to indices
        for dep_value in dep_values:
            dep_index = _normalize_dependency_index(
                dep_value, order_to_index_for_cycle, len(steps), node_index
            )
            if dep_index is not None and has_cycle(dep_index):
                return True
        
        rec_stack.remove(node_index)
        return False
    
    for i in range(len(steps)):
        if i not in visited and has_cycle(i):
            errors.append(f"Circular dependency detected involving step {i} ({steps[i].get('step_name', 'Unknown')})")
            break

    if coerced_values:
        logger.info(
            "[DependencyResolver] Coerced depends_on values to integers",
            extra={
                "coerced_count": len(coerced_values),
                "coerced_samples": coerced_values[:5],
                "total_steps": len(steps),
            },
        )
    
    return len(errors) == 0, errors


def get_ready_steps(completed_step_indices: List[int], all_steps: List[Dict]) -> List[int]:
    """
    Get steps that are ready to execute based on completed steps.
    
    Args:
        completed_step_indices: List of completed step indices
        all_steps: List of all step dictionaries
        
    Returns:
        List of ready step indices
    """
    completed = set(completed_step_indices)
    dependencies = build_dependency_graph(all_steps)
    ready_steps: List[int] = []
    
    for i in range(len(all_steps)):
        if i in completed:
            continue  # Already completed
        
        deps = dependencies.get(i, [])
        all_deps_completed = len(deps) == 0 or all(dep_index in completed for dep_index in deps)
        
        if all_deps_completed:
            ready_steps.append(i)
    
    return ready_steps


def get_step_status(
    completed_step_indices: List[int],
    running_step_indices: List[int],
    all_steps: List[Dict]
) -> Dict[int, str]:
    """
    Get step status for all steps.
    
    Args:
        completed_step_indices: List of completed step indices
        running_step_indices: List of running step indices
        all_steps: List of all step dictionaries
        
    Returns:
        Dictionary mapping step index to status string
    """
    status: Dict[int, str] = {}
    completed = set(completed_step_indices)
    running = set(running_step_indices)
    ready_steps = get_ready_steps(completed_step_indices, all_steps)
    
    for index in range(len(all_steps)):
        if index in completed:
            status[index] = StepStatus.COMPLETED.value
        elif index in running:
            status[index] = StepStatus.RUNNING.value
        elif index in ready_steps:
            status[index] = StepStatus.READY.value
        else:
            status[index] = StepStatus.WAITING.value
    
    return status

File: backend/api/src/utils/dependencyResolver.ts
/**
 * Dependency Resolution Engine
 * Handles building dependency graphs, detecting parallel opportunities, and resolving execution groups.
 *
 * This module provides algorithms for:
 * - Building dependency graphs from workflow steps
 * - Detecting opportunities for parallel execution
 * - Resolving execution groups for optimal step ordering
 * - Validating dependencies and detecting cycles
 *
 * @module dependencyResolver
 */

import {
  WorkflowStep,
  DependencyGraph,
  ExecutionGroup,
  ExecutionPlan,
  StepStatus,
  ValidationResult,
} from "./types";

/**
 * Build dependency graph from workflow steps.
 *
 * Creates a bidirectional dependency graph that tracks both:
 * - Dependencies: which steps each step depends on
 * - Dependents: which steps depend on each step
 *
 * Dependencies can be explicitly defined via `depends_on` (source of truth) or auto-detected from `step_order` (legacy fallback).
 *
 * @param steps - Array of workflow steps
 * @returns Dependency graph with dependencies and dependents maps
 * @throws {Error} If steps array is empty or invalid
 *
 * @example
 * ```typescript
 * const steps: WorkflowStep[] = [
 *   { step_name: 'Step 1', model: 'gpt-4', instructions: '...', step_order: 0 },
 *   { step_name: 'Step 2', model: 'gpt-4', instructions: '...', step_order: 1, depends_on: [0] },
 *   { step_name: 'Step 3', model: 'gpt-4', instructions: '...', step_order: 1, depends_on: [0] },
 * ];
 * const graph = buildDependencyGraph(steps);
 * // graph.dependencies: Map(0 => [], 1 => [0], 2 => [0])
 * // graph.dependents: Map(0 => [1, 2], 1 => [], 2 => [])
 * ```
 */
export function buildDependencyGraph(steps: WorkflowStep[]): DependencyGraph {
  if (!Array.isArray(steps) || steps.length === 0) {
    throw new Error("Steps must be a non-empty array");
  }
  const dependencies = new Map<number, number[]>();
  const dependents = new Map<number, number[]>();

  // Initialize maps
  steps.forEach((_, index) => {
    dependencies.set(index, []);
    dependents.set(index, []);
  });

  steps.forEach((step, index) => {
    if (step.depends_on && Array.isArray(step.depends_on)) {
      // Explicit dependencies provided
      const deps = step.depends_on.filter(
        (depIndex) =>
          depIndex >= 0 && depIndex < steps.length && depIndex !== index,
      );
      dependencies.set(index, deps);

      // Update dependents map
      deps.forEach((depIndex) => {
        const currentDependents = dependents.get(depIndex) || [];
        if (!currentDependents.includes(index)) {
          currentDependents.push(index);
          dependents.set(depIndex, currentDependents);
        }
      });
    } else {
      // Auto-detect from step_order
      const stepOrder = step.step_order !== undefined ? step.step_order : index;

      // Find all steps with lower step_order
      const lowerOrderSteps = steps
        .map((s, i) => ({
          step: s,
          index: i,
          order: s.step_order !== undefined ? s.step_order : i,
        }))
        .filter(({ order }) => order < stepOrder)
        .map(({ index }) => index);

      if (lowerOrderSteps.length > 0) {
        dependencies.set(index, lowerOrderSteps);

        // Update dependents map
        lowerOrderSteps.forEach((depIndex) => {
          const currentDependents = dependents.get(depIndex) || [];
          if (!currentDependents.includes(index)) {
            currentDependents.push(index);
            dependents.set(depIndex, currentDependents);
          }
        });
      }
    }
  });

  return {
    steps,
    dependencies,
    dependents,
  };
}

/**
 * Detect parallel opportunities from step_order.
 * Steps with the same step_order can potentially run in parallel.
 *
 * @param steps - Array of workflow steps
 * @returns Map of step_order to array of step indices with that order
 *
 * @example
 * ```typescript
 * const opportunities = detectParallelOpportunities(steps);
 * // If steps 0, 1, 2 all have step_order: 0, returns Map(0 => [0, 1, 2])
 * ```
 */
export function detectParallelOpportunities(
  steps: WorkflowStep[],
): Map<number, number[]> {
  const orderGroups = new Map<number, number[]>();

  steps.forEach((step, index) => {
    const stepOrder = step.step_order !== undefined ? step.step_order : index;

    if (!orderGroups.has(stepOrder)) {
      orderGroups.set(stepOrder, []);
    }
    orderGroups.get(stepOrder)!.push(index);
  });

  return orderGroups;
}

/**
 * Resolve execution groups - group steps into batches that can run in parallel.
 *
 * Uses topological sorting to determine the optimal execution order while maximizing
 * parallel execution opportunities. Steps in the same group can run simultaneously.
 *
 * @param steps - Array of workflow steps
 * @returns Execution plan with grouped steps
 *
 * @example
 * ```typescript
 * const plan = resolveExecutionGroups(steps);
 * // plan.executionGroups: [
 * //   { groupIndex: 0, stepIndices: [0], canRunInParallel: false },
 * //   { groupIndex: 1, stepIndices: [1, 2], canRunInParallel: true },
 * // ]
 * ```
 */
export function resolveExecutionGroups(steps: WorkflowStep[]): ExecutionPlan {
  if (steps.length === 0) {
    return {
      executionGroups: [],
      totalSteps: 0,
    };
  }

  const graph = buildDependencyGraph(steps);
  const executionGroups: ExecutionGroup[] = [];
  const completed = new Set<number>();
  let groupIndex = 0;

  while (completed.size < steps.length) {
    // Find all steps that are ready to execute (all dependencies completed)
    const readySteps: number[] = [];

    for (let i = 0; i < steps.length; i++) {
      if (completed.has(i)) {
        continue; // Already completed
      }

      const deps = graph.dependencies.get(i) || [];
      const allDepsCompleted =
        deps.length === 0 || deps.every((depIndex) => completed.has(depIndex));

      if (allDepsCompleted) {
        readySteps.push(i);
      }
    }

    if (readySteps.length === 0) {
      // This shouldn't happen if dependencies are valid, but handle gracefully
      console.warn(
        "No ready steps found, but not all steps completed. Possible circular dependency.",
      );
      break;
    }

    // Check if steps in this group can run in parallel
    // Steps can run in parallel if they don't depend on each other
    const canRunInParallel =
      readySteps.length > 1 && !hasInternalDependencies(readySteps, graph);

    executionGroups.push({
      groupIndex,
      stepIndices: readySteps,
      canRunInParallel: canRunInParallel || readySteps.length === 1,
    });

    // Mark these steps as completed for next iteration
    readySteps.forEach((stepIndex) => completed.add(stepIndex));
    groupIndex++;
  }

  return {
    executionGroups,
    totalSteps: steps.length,
  };
}

/**
 * Check if steps in a group have internal dependencies (can't run in parallel)
 */
function hasInternalDependencies(
  stepIndices: number[],
  graph: DependencyGraph,
): boolean {
  for (const stepIndex of stepIndices) {
    const deps = graph.dependencies.get(stepIndex) || [];
    // If any dependency is also in this group, they can't run in parallel
    if (deps.some((depIndex) => stepIndices.includes(depIndex))) {
      return true;
    }
  }
  return false;
}

/**
 * Validate dependencies - check for circular dependencies and invalid references.
 *
 * Performs comprehensive validation:
 * - Checks for invalid dependency indices (out of range, self-references)
 * - Detects circular dependencies using DFS
 * - Validates both explicit and auto-detected dependencies
 *
 * @param steps - Array of workflow steps to validate
 * @returns Validation result with errors if any
 *
 * @example
 * ```typescript
 * const validation = validateDependencies(steps);
 * if (!validation.valid) {
 *   console.error('Dependency errors:', validation.errors);
 * }
 * ```
 */
export function validateDependencies(steps: WorkflowStep[]): ValidationResult {
  const errors: string[] = [];

  if (!steps || steps.length === 0) {
    return { valid: true, errors: [] };
  }

  // Check for invalid dependency indices
  steps.forEach((step, index) => {
    if (step.depends_on && Array.isArray(step.depends_on)) {
      step.depends_on.forEach((depIndex: number) => {
        if (depIndex < 0 || depIndex >= steps.length) {
          errors.push(
            `Step ${index} (${step.step_name}): depends_on index ${depIndex} is out of range`,
          );
        }
        if (depIndex === index) {
          errors.push(
            `Step ${index} (${step.step_name}): cannot depend on itself`,
          );
        }
      });
    }
  });

  // Check for circular dependencies using DFS
  const visited = new Set<number>();
  const recStack = new Set<number>();

  function hasCycle(nodeIndex: number): boolean {
    if (recStack.has(nodeIndex)) {
      return true; // Found a cycle
    }
    if (visited.has(nodeIndex)) {
      return false; // Already processed
    }

    visited.add(nodeIndex);
    recStack.add(nodeIndex);

    const step = steps[nodeIndex];
    const deps = step.depends_on || [];

    // If no explicit dependencies, check step_order
    if (deps.length === 0 && step.step_order !== undefined) {
      const stepOrder = step.step_order;
      const lowerOrderSteps = steps
        .map((s, i) => ({
          step: s,
          index: i,
          order: s.step_order !== undefined ? s.step_order : i,
        }))
        .filter(({ order }) => order < stepOrder)
        .map(({ index }) => index);

      for (const depIndex of lowerOrderSteps) {
        if (hasCycle(depIndex)) {
          return true;
        }
      }
    } else {
      for (const depIndex of deps) {
        if (hasCycle(depIndex)) {
          return true;
        }
      }
    }

    recStack.delete(nodeIndex);
    return false;
  }

  for (let i = 0; i < steps.length; i++) {
    if (!visited.has(i) && hasCycle(i)) {
      errors.push(
        `Circular dependency detected involving step ${i} (${steps[i].step_name})`,
      );
      break;
    }
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Get steps that are ready to execute based on completed steps.
 *
 * A step is ready when all of its dependencies have been completed.
 *
 * @param completedStepIndices - Array of indices of completed steps
 * @param allSteps - All workflow steps
 * @returns Array of step indices that are ready to execute
 *
 * @example
 * ```typescript
 * const readySteps = getReadySteps([0], allSteps);
 * // Returns steps that only depend on step 0
 * ```
 */
export function getReadySteps(
  completedStepIndices: number[],
  allSteps: WorkflowStep[],
): number[] {
  const completed = new Set(completedStepIndices);
  const graph = buildDependencyGraph(allSteps);
  const readySteps: number[] = [];

  for (let i = 0; i < allSteps.length; i++) {
    if (completed.has(i)) {
      continue; // Already completed
    }

    const deps = graph.dependencies.get(i) || [];
    const allDepsCompleted =
      deps.length === 0 || deps.every((depIndex) => completed.has(depIndex));

    if (allDepsCompleted) {
      readySteps.push(i);
    }
  }

  return readySteps;
}

/**
 * Get step status for all steps.
 *
 * Determines the current status of each step:
 * - 'completed': Step has finished execution
 * - 'running': Step is currently executing
 * - 'ready': Step is ready to execute (all dependencies completed)
 * - 'waiting': Step is waiting for dependencies to complete
 *
 * @param completedStepIndices - Array of indices of completed steps
 * @param runningStepIndices - Array of indices of currently running steps
 * @param allSteps - All workflow steps
 * @returns Map of step index to status
 *
 * @example
 * ```typescript
 * const status = getStepStatus([0], [1], allSteps);
 * // status.get(0): 'completed'
 * // status.get(1): 'running'
 * // status.get(2): 'ready' or 'waiting' depending on dependencies
 * ```
 */
export function getStepStatus(
  completedStepIndices: number[],
  runningStepIndices: number[],
  allSteps: WorkflowStep[],
): Map<number, StepStatus> {
  const status = new Map<number, StepStatus>();
  const completed = new Set(completedStepIndices);
  const running = new Set(runningStepIndices);
  const readySteps = getReadySteps(completedStepIndices, allSteps);

  allSteps.forEach((_, index) => {
    if (completed.has(index)) {
      status.set(index, "completed");
    } else if (running.has(index)) {
      status.set(index, "running");
    } else if (readySteps.includes(index)) {
      status.set(index, "ready");
    } else {
      status.set(index, "waiting");
    }
  });

  return status;
}
